application/src/main/java/org/togetherjava/tjbot/commands/help/BookmarksPaginatorInteractor.java

```diff
@@ -0,0 +1,489 @@
+package org.togetherjava.tjbot.commands.help;
+
+import net.dv8tion.jda.api.EmbedBuilder;
+import net.dv8tion.jda.api.JDA;
+import net.dv8tion.jda.api.entities.MessageEmbed;
+import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel;
+import net.dv8tion.jda.api.entities.emoji.Emoji;
+import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent;
+import net.dv8tion.jda.api.events.interaction.command.GenericCommandInteractionEvent;
+import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
+import net.dv8tion.jda.api.events.interaction.component.SelectMenuInteractionEvent;
+import net.dv8tion.jda.api.interactions.components.ActionRow;
+import net.dv8tion.jda.api.interactions.components.ComponentInteraction;
+import net.dv8tion.jda.api.interactions.components.LayoutComponent;
+import net.dv8tion.jda.api.interactions.components.buttons.Button;
+import net.dv8tion.jda.api.interactions.components.selections.SelectMenu;
+import net.dv8tion.jda.api.interactions.components.selections.SelectOption;
+
+import org.togetherjava.tjbot.commands.UserInteractionType;
+import org.togetherjava.tjbot.commands.UserInteractor;
+import org.togetherjava.tjbot.commands.componentids.ComponentIdGenerator;
+import org.togetherjava.tjbot.commands.componentids.ComponentIdInteractor;
+import org.togetherjava.tjbot.commands.utils.Arraylizable;
+import org.togetherjava.tjbot.db.generated.tables.records.BookmarksRecord;
+
+import java.awt.*;
+import java.util.*;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+
+/**
+ * This class can create a paginated message for listing or removing bookmarks. A paginated message
+ * is requested by calling
+ * {@link BookmarksSystem#requestListPagination(GenericCommandInteractionEvent)} or
+ * {@link BookmarksSystem#requestRemovePagination(GenericCommandInteractionEvent)}. This class then
+ * generates a page embed with navigation components and if it is a remove page also the remove
+ * components. Then the paginator waits for interactions with the components. The action for
+ * interacting with the component will be taken followed by updating the page embed and components.
+ */
+public final class BookmarksPaginatorInteractor implements UserInteractor {
```

Comment:
Okay. So before I dive deeper into the actual review of the logic in the paginator, lets have a CR round were we fix the issues with the overall design. The main issue I have with this is how you defined the responsibilities and the flow. Right now its very complex, heavily wired and a tad messy.

Right now, you have:
1. `BookmarksSystem` knows who can do pagination
2. But it doesnt know it right from the start, but via late-init (`accept...`)
3. `BookmarksPaginatorInteractor` tells the system that he can do it (`accept...`), during its constructor call
4. then, the actual user `BookmarksCommand` forwards everything `list` and `remove` to the paginator via asking the system

Problems I have with this:
1. It is not `BookmarksSystem`s responsibility to do anything pagination related
2. It makes it overly complex that `BookmarksSystem` acts as some sort of middle-man here
3. `BookmarksPaginatorInteractor` should not expose itself during its own constructor call
4. The wiring between the system and paginator being late-init feels like u attempted to design it as some sort of general purpose solution, but it is a very specific solution that is only meant to be used for the bookmark feature. So instead, you should just wire it directly. The classes are close to each other and know each other. The system could just directly use the paginator without that `accept` stuff
5. It is not the paginators responsibility to do list or remove stuff. Only to do pagination. Its the `BookmarksCommand`s responsibility to do that
6. It is clever that u made the paginator its own `UserFeature`, but this makes it actually quite confusing. Where would someone look when searching the entry point of clicking the delete button? Obviously they would look for a `onButtonClick` in `BookmarksCommand`, but wont find any. The connection is not obvious at all.

Alright, thats a lot of stuff to cover. So how to approach this? Id suggest the following alternative design:
1. Move all the pagination stuff into the `BookmarksCommand` first. U will probably get rid of a lot of general-purpose-utility and fluff u created to make this balancing act between all the classes.
2. Once done with that, we can look into moving some of the stuff into a delegate class. The main point here will be that `BookmarksCommand` still has its event entry-points like `onButtonClick` and `onSelectionMenu` etc. But will then just forward some stuff to this delegate.
3. This delegate would then be created, maintained and known only by `BookmarksCommand`, no other classes should use or know this. The delegate also doesnt need to be `UserFeature` anymore or anything

So all in all, start by getting rid of ur paginator class again, completely. And then we move some stuff into a delegate in the next step.

Reply:
Oh god that sounds like a lot of work. I have a good amount of school stuff to do so this could take a while. You can also try moving it yourself if you want to.

Reply:
sure, ill put it on my todo list. no worries 👍

Reply:
Great. I should be able to start working again on Friday